src/OpenLoco/src/Map/TileManager.cpp

```diff
@@ -464,16 +464,52 @@ namespace OpenLoco::World::TileManager
         return height;
     }
 
-    SmallZ getSurfaceCornerHeight(const SurfaceElement& surface)
+    static uint8_t getCornerDownMask(const uint8_t cornerUp)
+    {
+        switch (cornerUp)
+        {
+            case SurfaceSlope::CornerUp::north:
+                return SurfaceSlope::CornerDown::south;
+
+            case SurfaceSlope::CornerUp::south:
+                return SurfaceSlope::CornerDown::north;
+
+            case SurfaceSlope::CornerUp::west:
+                return SurfaceSlope::CornerDown::east;
+
+            case SurfaceSlope::CornerUp::east:
+                return SurfaceSlope::CornerDown::west;
+        }
+        return SurfaceSlope::flat;
+    }
+
+    // TODO: rename this function
+    SmallZ getSurfaceCornerHeight(const SurfaceElement& surface, const uint8_t cornerMask)
     {
         auto baseZ = surface.baseZ();
-        if (surface.slope())
+        if (surface.slope() & cornerMask)
         {
             baseZ += kSmallZStep;
+            uint8_t cornerDownMask = getCornerDownMask(cornerMask);
+            if (surface.isSlopeDoubleHeight() && surface.slopeCorners() == cornerDownMask)
+            {
+                baseZ += kSmallZStep;
+            }
         }
-        if (surface.isSlopeDoubleHeight())
+
+        return baseZ;
+    }
+
+    SmallZ getSurfaceCornerHeight(const SurfaceElement& surface)
+    {
+        auto baseZ = surface.baseZ();
```

Comment:
this method is really similar to the one above, is there any way to call that one with something like 0xFF as the cornerMask? would like to reduce duplicate code, that's all

Reply:
This is the original method; the other is quite different in that it queries the opposite corners. I originally "optimized" this by doing the same thing you suggested, but that meant that I did not query the `cornerDownMask` instead.

They really ought to be two separate functions. This is why I placed a comment suggesting that a different name for each might be better... But I haven't come up with anything. Suggestions?

Reply:
I thought they are really doing the same thing though:

```
SmallZ getSurfaceCornerHeight(const SurfaceElement& surface, const uint8_t cornerUp)
{
    auto baseZ = surface.baseZ();
    if (surface.slope() & cornerUp)
    {
        baseZ += kSmallZStep;
        uint8_t cornerDownMask = getCornerDownMask(cornerUp);
        if (surface.isSlopeDoubleHeight() && (cornerUp == 0xFF || surface.slopeCorners() == cornerDownMask))
        {
            baseZ += kSmallZStep;
        }
    }

    return baseZ;
}

SmallZ getSurfaceCornerHeight(const SurfaceElement& surface)
{
    return getSurfaceCornerHeight(surface, 0xFF);
}
```

Reply:
That is more or less what I initially did: https://github.com/OpenLoco/OpenLoco/pull/2167/commits/67294040a1ea2f476e3b2e55a568df5ac6d42e32, which then turned into the current code in https://github.com/OpenLoco/OpenLoco/pull/2167/commits/4dd609ff33022bf44c24b5c5417a7d8b88267e12. You're right that there's some overlap, but due to the added corner down check, I think this would best remain two functions.

Reply:
fair, happy to leave it with that you've got then. i can't think of any good method names at the moment